KAFKA-5379 - ProcessorContext.appConfigs() should return parsed values.

twbecker opened the pull request.

twbecker committed:
KAFKA-5379 - ProcessorContext.appConfigs() should return parsed values.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5769/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5755/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5770/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5756/
Test PASSed (JDK 8 and Scala 2.12).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java:
```diff
@@ -192,6 +192,10 @@
      * The config properties are defined in the {@link org.apache.kafka.streams.StreamsConfig}
      * object and associated to the ProcessorContext.
+     * <p>
+     * The type of the values is dependent on the {@link org.apache.kafka.common.config.ConfigDef.Type} of the property
```
Nit: `{@link org.apache.kafka.common.config.ConfigDef.Type type}` similar below.

twbecker replied:
Done.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:
```diff
@@ -156,7 +157,10 @@ public long timestamp() {
     @Override
     public Map<String, Object> appConfigs() {
-        return config.originals();
+        Map<String, Object> combined = new HashMap<>();
```
nit: add `final`

twbecker replied:
Done.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContextTest.java:
```diff
@@ -136,10 +139,28 @@ public void shouldReturnTimestampFromRecordContext() throws Exception {
+        static Properties config;
+        static {
+            config = minimalStreamsConfig();
+            config.put(StreamsConfig.ROCKSDB_CONFIG_SETTER_CLASS_CONFIG, RocksDBConfigSetter.class.getName());
```
Add a comment that this must be a String type to make a useful test.

twbecker replied:
Done.

mjsax reviewed (changes requested):
Can you also fix `RocksDBStore#openDB()` to use this new functionality to avoid the parsing step there. Thx.

Commit:
Address review comments for KAFKA-5379.

twbecker commented:
Updated `RocksDBStore` as you suggested. I also updated the test to merely ensure the `RocksDBConfigSetter` is called, since the String vs Class config variations are now tested elsewhere.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5783/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5797/
Test FAILed (JDK 7 and Scala 2.11).

mjsax reviewed (approved):
LGTM.

Call for second review @enothereska @bbejeck

mjsax commented:
Unclear why test failed. But it was in core, so unrelated.

mjsax commented:
Retest this please.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5815/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5801/
Test FAILed (JDK 8 and Scala 2.12).

mjsax commented:
Test failed due to environmental issue...

mjsax commented:
Retest this please.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5819/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5805/
Test PASSed (JDK 8 and Scala 2.12).

twbecker committed:
Merge branch 'trunk' into kafka-5379

mjsax commented:
@twbecker There are checkstyle errors in the build. Will review after those got fixed. Thx.

twbecker commented:
Yeah, that's what I get for fixing the conflicts in the Github web UI ;)

guozhangwang commented:
retest this please

guozhangwang commented:
retest this please

guozhangwang commented:
retest this please

guozhangwang commented:
@twbecker The jenkins failures are related:

```
:streams:checkstyleMain[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11@2/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:20:8: Unused import - org.apache.kafka.common.config.ConfigDef. [UnusedImports]
 FAILED
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreTest.java:
```diff
@@ -338,22 +333,42 @@ public void setConfig(final String storeName, final Options options, final Map<S
+    private static class ConfigurableProcessorContext extends AbstractProcessorContext {
+        private final File stateDir;
+
+        ConfigurableProcessorContext(StreamsConfig config, File stateDir) {
```
Nit: add `final` to both parameters -- please follow a "one parameter per line" formatting.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreTest.java:
```diff
@@ -338,22 +333,42 @@ public void setConfig(final String storeName, final Options options, final Map<S
+        }
+
+        @Override
+        public StateStore getStateStore(String name) {
```
Nit: `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreTest.java:
```diff
@@ -338,22 +333,42 @@ public void setConfig(final String storeName, final Options options, final Map<S
         @Override
-        public Map<String, Object> appConfigs() {
-            return configs;
+        public Cancellable schedule(long interval, PunctuationType type, Punctuator callback) {
```
Nit: `final` and formatting

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreTest.java:
```diff
@@ -338,22 +333,42 @@ public void setConfig(final String storeName, final Options options, final Map<S
-                                     final Map<String, Object> configs) {
-            super(stateDir, keySerde, valSerde, collector, cache);
-            this.configs = configs;
+    private static class ConfigurableProcessorContext extends AbstractProcessorContext {
```
Why not stay with `MockProcessorContext` -- this should avoid to overwrite all the empty methods below?

twbecker replied:
It's been a while since I did this, but I think it was because I needed to specify the `StreamsConfig`.

guozhangwang replied:
I think we can do it in a follow-up PR after merging this one.

mjsax reviewed (changes requested):
Thanks for the PR @twbecker !

Commit:
Address review comments

twbecker commented:
Ok, @mjsax should be gtg now.

guozhangwang commented:
Merged to trunk. Thanks @twbecker .

mjsax commented:
Thanks @twbecker !
